dts/bindings/misc/raspberrypi,pico-pio-device.yaml

```diff
@@ -1,9 +1,17 @@
+# Copyright (c) 2023 Ionut Pavel <iocapa@iocapa.com>
 # Copyright (c) 2023, TOKITA Hiroshi
 # Copyright (c) 2023, Yonatan Schachter
 # SPDX-License-Identifier: Apache-2.0
 
 description: Raspberry Pi Pico PIO device
 
-compatible: "raspberrypi,pico-pio-device"
+on-bus: rpi-pico-pio
```

Comment:
In my understanding, this PR contains

1. Addition of interrupt handling
2. PIO program macro
3. Introduction of buses
4. Adding samples

I think 2 and 3 are easily acceptable, so you might want to raise them as separate PRs.
(In other words, 1 is probably  controversial.)

Reply:
@soburi 

This PR contains a different implementation for handling the PIO / PIO devices.
**There is no API compatibility with the current implementation**, therefore it cannot be split, as builds will fail.

Aside from some similarities on the CMake/Kconfig/naming side, pretty much everything on the source side is a different design, see:

The "PIO" driver, contains (shared) instructions / SM allocators and shared interrupts handling.
[pio_rpi_pico.c](https://github.com/zephyrproject-rtos/zephyr/blob/293f573201218758d64b7952bc9a24752841794b/drivers/misc/pio_rpi_pico/pio_rpi_pico.c)
[pio_rpi_pico.h](https://github.com/zephyrproject-rtos/zephyr/blob/293f573201218758d64b7952bc9a24752841794b/include/zephyr/drivers/misc/pio_rpi_pico/pio_rpi_pico.h)

This is a full(ish) flegded uart over PIO implementation supporting multiple data/stop bits configurations and interrupt driven mode.
[uart_rpi_pico_pio.c](https://github.com/zephyrproject-rtos/zephyr/blob/293f573201218758d64b7952bc9a24752841794b/drivers/serial/uart_rpi_pico_pio.c)